### PR TITLE
Enhance simplifier by adding Canonicalize

### LINKED ISSUE #8724: Expression simplifier does not simplify `A = B AND B = A`
### Is your feature request related to a problem or challenge?

Given this case:

```
DataFusion CLI v34.0.0
❯ CREATE TABLE kumachan (wakana string) AS VALUES ('ookuma');
0 rows in set. Query took 0.004 seconds.

❯ explain select * from kumachan where wakana = 'ookuma' and 'ookuma' = wakana;
+---------------+-------------------------------------------------------------------------------+
| plan_type     | plan                                                                          |
+---------------+-------------------------------------------------------------------------------+
| logical_plan  | Filter: kumachan.wakana = Utf8("ookuma") AND Utf8("ookuma") = kumachan.wakana |
|               |   TableScan: kumachan projection=[wakana]                                     |
| physical_plan | CoalesceBatchesExec: target_batch_size=8192                                   |
|               |   FilterExec: wakana@0 = ookuma AND ookuma = wakana@0                         |
|               |     MemoryExec: partitions=1, partition_sizes=[1]                             |
|               |                                                                               |
+---------------+-------------------------------------------------------------------------------+
2 rows in set. Query took 0.004 seconds.

❯
```

I would expect the final plan to simplify the `FilterExec` to only be `wakana@0 = ookuma` or `ookuma = wakana@0`, as the `AND` of these conditions is redundant.

### Describe the solution you'd like

Possible solutions:

1. introduce new optimizer rule to reorder the `left` and `right` fields of all `BinaryExpr` with `op` of `Operator::Eq` to a consistent order, which would be placed high in the rules list (during analysis?) to ensure downstream rules (like simplify_expressions) can properly determine this equality and simplify the condition
2. modify `BinaryExpr` to have a custom `PartialEq`/`Eq` implementation which disregards order of its `left`/`right` fields when checking equality (only for `Operator::eq`)

https://github.com/apache/arrow-datafusion/blob/9a6cc889a40e4740bfc859557a9ca9c8d043891e/datafusion/expr/src/expr.rs#L208-L217

Option 1 seems the cleaner solution, since don't have to muck with manual implementation of `PartialEq`/`Eq`

### Describe alternatives you've considered

_No response_

### Additional context

Found this while checking if #4732 can be closed

### PR DESCRIPTION
**Pull request by yyy1000**
## Which issue does this PR close?

Closes #8724 

## Rationale for this change

introduce new optimizer rule and optimizer can be better.

## What changes are included in this PR?

1. Reorder all BinaryExpr to <col> <op> <literal> or <col1> <op> <col2> (the name of col1 is larger than col2)
2. two unit test cases

## Are these changes tested?
Yes

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 1]
```diff
@@ -224,6 +226,63 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
...
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+        if let Expr::BinaryExpr(BinaryExpr { left, op, right }) = expr {
+            // Case 1, <col1> <op> <col2>
+            let mut new_expr = BinaryExpr {
+                left: left.clone(),
+                op: op.clone(),
+                right: right.clone(),
+            };
+            let mut switch_op: Operator = op.clone();
+            if left.try_into_col().is_ok() && right.try_into_col().is_ok() {
```
Comment (Jefffrey) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:254]: I think this check is better off done via a match, e.g.

```rust
match (left.as_ref(), right.as_ref()) {
    (Expr::Column(a), Expr::Column(b)) => todo!(),
    _ => todo!(),
}
```

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 2]
```diff
@@ -224,6 +226,63 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
...
+                let left_name = left.canonical_name();
+                let right_name = right.canonical_name();
+                if left_name < right_name {
+                    if let Some(swap_op) = op.swap() {
+                        switch_op = swap_op;
+                    }
```
Comment (Jefffrey) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:260]: You're on the right track here with swapping the Operator, but one very important note is that for this functionality, we should only swap certain operators that it makes sense to.

For example, we cannot swap `A - B` to `B - A`, as that changes the expression's meaning

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 3]
```diff
@@ -224,6 +226,63 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
...
+                    new_expr = BinaryExpr {
+                        left: right,
+                        op: switch_op,
+                        right: left,
+                    };
+                }
+            }
+            // Case 2, <literal> <op> <col>
+            else if left.try_into_col().is_err() && right.try_into_col().is_ok() {
```
Comment (Jefffrey) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:269]: One advantage of using `match` to destructure the `left`/`right` `Expr`s is that you would avoid cases like this, because a pitfall here is that just because an `Expr` is not a `Expr::Column` variant, doesn't mean it is a `Expr::Literal` variant, so failing to cast `left` to a `Column` could mean its another variant than `Expr::Literal`

**Review by Jefffrey — CHANGES_REQUESTED** [review]
You're thinking in the right direction, but there'll need to be more work done to achieve the intended functionality. I've left some pointers to hopefully help you adjust the code as needed, but feel free to ask if you need further clarification or help

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs (hunk 2 above)
Comment (yyy1000) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:260]: Got it.
Here `swap` would return None if operator makes no sense, e.g.: `-`, see: https://github.com/apache/arrow-datafusion/blob/cc4289484c33e478242bf5d2b59f695fdb427ab9/datafusion/expr/src/operator.rs#L172C5-L202C6

I think in this case we will not change the order of the Expr, and <col1> and <col2> would remain the same order as before.

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs (hunk 1 above)
Comment (yyy1000) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:254]: Got it!
Thank you very much

**Comment by yyy1000** [comment]
Changed the code according to the review. :)

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 4]
```diff
@@ -224,6 +226,62 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
     }
 }
 
+#[allow(rustdoc::private_intra_doc_links)]
+/// Canonicalize any BinaryExprs that are not in canonical form
+/// <literal> <op> <col> is rewritten to <col> <op> <literal> (remember to switch the operator)
+/// <col> <op> <literal> is canonical
+/// <col1> <op> <col2> is rewritten so that the name of col1 sorts higher than col2 (b > a would be canonicalized to a < b)
```
Comment (Jefffrey) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:233]: Might need to upgrade the docs here, for example remove the `remember to switch the operator` line (or change it to clarify that operator would be switched)

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 5]
```diff
@@ -224,6 +226,62 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
...
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+        if let Expr::BinaryExpr(BinaryExpr { left, op, right }) = expr {
+            // Case 1, <col1> <op> <col2>
+            let mut new_expr = BinaryExpr {
+                left: left.clone(),
+                op: op.clone(),
+                right: right.clone(),
+            };
+            match (left.as_ref(), right.as_ref()) {
```
Comment (Jefffrey) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:253]: You could even push the `op.swap()` into this match, to remove the nested if blocks, e.g.

```rust
match (left.as_ref(), right.as_ref(), op.swap()) {
    (Expr::Column(_), Expr::Column(_), Some(swapped_op)) => {
        todo!(); // swap based on name, directly using the swapped_op
    }
    (Expr::Literal(_), Expr::Column(_), Some(swapped_op)) => {
        todo!(); // swap to have column on left
    }
    _ => todo!(),
}
```

You could then even add the name check of left vs right as an if condition to the branch

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 6]
```diff
@@ -224,6 +226,62 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
...
+    type N = Expr;
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+        if let Expr::BinaryExpr(BinaryExpr { left, op, right }) = expr {
+            // Case 1, <col1> <op> <col2>
+            let mut new_expr = BinaryExpr {
+                left: left.clone(),
+                op: op.clone(),
+                right: right.clone(),
+            };
```
Comment (Jefffrey) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:252]: I wonder if there is a cleaner way to do this, without having to clone here (since this will always clone for `Expr;:BinaryExpr` even if we don't end up swapping)

Could try an approach without relying on mutability of `new_expr`

**Review by Jefffrey — COMMENTED** [review]
Looking better! :+1: 

I *think* the logic is good, but will see if CI passes to ensure no regression in behaviour.

I've left some more comments, mainly around the structure of the code and how to make it more ergonomic in a Rust way (in my opinion, at least)

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs (hunk 6 above)
Comment (yyy1000) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs]: I use `return` in `if` statement to avoid `new_expr` in the new changes :)

**Comment by yyy1000** [comment]
Thank you for your detailed review @Jefffrey !
I changed the code, and now it looks better :)
I'm new to Rust so the code may not looks good at first, thanks again for your guidance.

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 7]
```diff
@@ -224,6 +226,51 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
...
+        Self {}
+    }
+}
+
+impl TreeNodeRewriter for Canonicalizer {
+    type N = Expr;
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+        if let Expr::BinaryExpr(BinaryExpr { left, op, right }) = expr {
+            match (left.as_ref(), right.as_ref(), op.swap()) {
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:246]: You can do something like this to reduce the indent of this function if you want (not required, I just figured I would point it out given you say you are looking to improve your rust skills

```suggestion
        let Expr::BinaryExpr(BinaryExpr { left, op, right }) = expr else {
          return Ok(expr)
        };
        match (left.as_ref(), right.as_ref(), op.swap()) {
```

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 8]
```diff
@@ -1594,6 +1641,20 @@ mod tests {
...
+            let expr = lit(1).lt(col("c2")).and(col("c2").gt(lit(1)));
+            let expected = col("c2").gt(lit(1));
+            assert_eq!(simplify(expr), expected);
+        }
+        {
+            let expr = col("c1").lt(col("c2")).and(col("c2").gt(col("c1")));
+            let expected = col("c2").gt(col("c1"));
+            assert_eq!(simplify(expr), expected);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (yyy1000) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1658]: Correct, in the test case, for `A < 5 AND A >= 5 --> same`, it will make no change :)

**Comment by yyy1000** [comment]
Thank you for the review @alamb !
I pushed the change to the PR and I will look into the CI failure if it's there :)

**Comment by alamb** [comment]
I kicked off the CI checks!

**Comment by yyy1000** [comment]
There's an `FieldNotFound` error in tests introduced by the PR. 🧐
I will dive into it and find what's wrong with it. :)

**Comment by yyy1000** [comment]
I figured it as a `SchemaError :: FieldNotFound` error and it seems a little complicated for me :(
Any guidance would be helpful.

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 9]
```diff
@@ -224,6 +226,52 @@ impl<S: SimplifyInfo> ExprSimplifier<S> {
...
+            (Expr::Literal(_a), Expr::Column(_b), Some(swapped_op)) => {
+                return Ok(Expr::BinaryExpr(BinaryExpr {
+                    left: right,
+                    op: swapped_op,
+                    right: left,
+                }));
+            }
+            _ => {}
+        }
+        Ok(Expr::BinaryExpr(BinaryExpr { left, op, right }))
```
Comment (Jefffrey) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:271]:
```suggestion
        match (left.as_ref(), right.as_ref(), op.swap()) {
            // <col1> <op> <col2>
            (Expr::Column(left_col), Expr::Column(right_col), Some(swapped_op)) if right_col > left_col => {
                Ok(Expr::BinaryExpr(BinaryExpr {
                    left: right,
                    op: swapped_op,
                    right: left,
                }))
            }
            // <literal> <op> <col>
            (Expr::Literal(_a), Expr::Column(_b), Some(swapped_op)) => {
                Ok(Expr::BinaryExpr(BinaryExpr {
                    left: right,
                    op: swapped_op,
                    right: left,
                }))
            }
            _ => Ok(Expr::BinaryExpr(BinaryExpr { left, op, right })),
        }
```

One last suggestion for code structure, to take better advantage of how Rust enables expressions to simplify things

Essentially pushing the if condition into that first match arm, and then removing the need to have a `return` keyword as all the match arms will have implicit return

**Review by Jefffrey — COMMENTED** [review]
I'll try take a look at the failing test and see if I can help explain/debug :+1:

**Comment by yyy1000** [comment]
Woo, much appreciated that @Jefffrey ❤️
I learned that using `explain` is a good way to debug, will see the methods to fix it now. :)

**Comment by yyy1000** [comment]
As I looked into the code today, 
if we respect the current behaviour, maybe we can change the `info` field in `ExprSimplifier`, and  see whether `info` contains 'Join infomation' https://github.com/apache/arrow-datafusion/blob/d9a1d4261f78dfd3ead235d6b850c64e0ae9bec6/datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs#L132-L148
if trying to fix that to not rely on join order, would it be implementable?

**Comment by yyy1000** [comment]
Woops, I found implement the second way would be more easy.
I pushed my changes to the PR, it would try make `column a` the right child and `column b` the right child if it can't find them in `left and right` child.
Also, two test cases are modified because the name of `t2.c1` is larger than` t1.c1`, and it meets expectation.

**Comment by alamb** [comment]
Retriggered CI

**Comment by yyy1000** [comment]
Ah, I believe the CI fails just because the test cases need to be changed due to `bigger names would be in the front`, I could do that then.

**Comment by yyy1000** [comment]
Well, I tried to fix the CI-error, but it's harder than I thought. :(
Now the problem is `on`, the two columns in Join, would change but some functions still pass left and right infos, didn't knowing that now the two columns in `on` has been swapped.
The failed test case ,

>  index out of bounds: the len is 2 but the index is 2 

indicated that here is the reason, https://github.com/apache/arrow-datafusion/blob/eb81ea299aa7e121bbe244e7e1ab56513d4ef800/datafusion/physical-plan/src/joins/utils.rs#L688-L697

**Comment by alamb** [comment]
🤔  so this PR has turned out to be far more complicated than we imagined

What is the current state of this PR? Do we think the code is ok, and we just need to update the tests? Or do we think something more fundamental is wrong with it?

Thank you for this work @yyy1000  -- I am sorry it was not an easy first issue as I had thought

**Comment by Jefffrey** [comment]
From what I understand, it seems this effort will be blocked as either have to:

- Fix left/right join assumptions throughout the code base
- Or enhance the expression simplifier to be able to tell when an Expr is from a join on clause, and if so don't do the reordering

**Comment by yyy1000** [comment]
Yeah, what @Jefffrey said is right.
I tried the first method but failed, now I'm trying the second method. It's kind of difficult for me. 🥲
Sorry for the delay. :(

**Comment by alamb** [comment]
> Yeah, what @Jefffrey said is right. I tried the first method but failed, now I'm trying the second method. It's kind of difficult for me. 🥲 Sorry for the delay. :(

No  worries -- thank you for your perseverence

File: datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs [hunk 10]
```diff
@@ -93,16 +93,31 @@ impl SimplifyExpressions {
...
+                .into_iter()
+                .map(|e| {
+                    // TODO: unify with `rewrite_preserving_name`
+                    let original_name = e.name_for_alias()?;
+                    let new_e = simplifier.simplify(e)?;
+                    new_e.alias_if_changed(original_name)
+                }).collect::<Result<Vec<_>>>()?
+            },
+            _ => {
+                plan
```
Comment (yyy1000) [inline datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs:109]: This is what I tried for the second method, I define a new function called `canonicalize` in `simplifier`.
And the rule `SimplifyExpressions` will call this if the `plan` is not `Join`.
I didn't change the current `simplify` function because I think add a param to judge whether the `Expr` is from a `Join` plan would break a lot.

**Review by yyy1000 — COMMENTED** [review]
Hi, here is what I tried the second method, now the `Join` would not call `canonicalize`.

File: datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs (hunk 10 above)
Comment (Jefffrey) [reply datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs]: Makes sense. Could you also add a comment to the code here explaining this? Otherwise its not immediately obvious for the near duplication of code :+1:

Comment (yyy1000) [reply datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs]: Sure!

**Comment by yyy1000** [comment]
Ah, I see the CI failed still due to some test cases need to be changed.
Would following the CI flow https://github.com/apache/arrow-datafusion/blob/903ef94a4742a4ff6933dcf696a287972df3b5e2/.github/workflows/rust.yml#L210-L239 to see the failure in my local environment a good method?

**Comment by yyy1000** [comment]
Update: following that helps me fix the wrong test cases. :)

**Comment by yyy1000** [comment]
@alamb I think this PR is close to be merged. :)
I'd like to see your review and feedback. 😃

File: datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs [hunk 11]
```diff
@@ -93,16 +93,33 @@ impl SimplifyExpressions {
...
-            .map(|e| {
-                // TODO: unify with `rewrite_preserving_name`
-                let original_name = e.name_for_alias()?;
-                let new_e = simplifier.simplify(e)?;
-                new_e.alias_if_changed(original_name)
-            })
-            .collect::<Result<Vec<_>>>()?;
+        let expr = match plan {
+            // Canonicalize step won't reorder expressions in a Join on clause.
+            // The left and right expressions in a Join on clause are not commutative.
```
Comment (Jefffrey) [inline datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs:98]:
```suggestion
            // Canonicalize step won't reorder expressions in a Join on clause.
            // The left and right expressions in a Join on clause are not commutative,
            // since the order of the columns must match the order of the children.
```

Just an extra bit of clarification

File: datafusion/physical-plan/src/joins/utils.rs [hunk 12]
```diff
@@ -293,9 +293,17 @@ fn check_join_set_is_valid(
...
+        let left_missing_reverse =
+            on_left_reverse.difference(left).collect::<HashSet<_>>();
+        let on_right_reverse = &on.iter().map(|on| on.0.clone()).collect::<HashSet<_>>();
+        let right_missing_reverse =
+            on_right_reverse.difference(right).collect::<HashSet<_>>();
+        if !left_missing_reverse.is_empty() | !right_missing_reverse.is_empty() {
+            return plan_err!(
+                "The left or right side of the join does not have all columns on \"on\": \nMissing on the left: {left_missing:?}\nMissing on the right: {right_missing:?}"
+            );
+        }
```
Comment (Jefffrey) [inline datafusion/physical-plan/src/joins/utils.rs:306]: Is this change still required if the order of the `on` clause in Joins should not be reordered, now?

**Comment by alamb** [comment]
I retriggered CI and will try and check it out later today

File: datafusion/physical-plan/src/joins/utils.rs (hunk 12 above)
Comment (yyy1000) [reply datafusion/physical-plan/src/joins/utils.rs:306]: Ah, got it. I should remove this.

**Review by Jefffrey — APPROVED** [review]
Tests passing, nice work :+1:

**Review by alamb — APPROVED** [review]
Thank you for sticking with this PR @yyy1000  and to @Jefffrey  for the assist and review 🙏

**Comment by alamb** [comment]
I merged this branch up from main to make sure all the CI passes, and if it does I plan to merge it in!

**Comment by alamb** [comment]
While reviewing this PR, I had a thought for a slightly more consistent API which I have proposed in https://github.com/apache/arrow-datafusion/pull/8983
